### ISSUE TITLE
Occasional hang in `tests/test_send_recv_obj.py`

### ISSUE THREAD

**Issue by TomAugspurger** [rapidsai/ucx-py#69]
```
pip install pytest-repeat
pytest -vs tests/test_send_recv_obj.py::test_send_recv_numpy --count=100
```

will usually reproduce.

Under UCX_LOG_LEVEL=DEBUG, I don't really notice a much of a difference between success and hangs (these are for send_recv_numba):  https://gist.github.com/0eaa7914ca6bf35bc2be88a4424e1b19

On the successful one, we have

```
  [1552583115.207903] [dgx05:35955:0]          ud_ep.c:1252 UCX  DEBUG ep 0x55c24154bd80: disconnect
  [1552583115.208373] [dgx05:35955:0]          flush.c:242  UCX  DEBUG close ep 0x7f92e490b048
  [1552583115.208382] [dgx05:35955:0]         ucp_ep.c:646  UCX  DEBUG ep 0x7f92e490b048: destroy
  [1552583115.208384] [dgx05:35955:0]         ucp_ep.c:656  UCX  DEBUG ep 0x7f92e490b048: cleanup lanes
  [1552583115.208386] [dgx05:35955:0]         ucp_ep.c:661  UCX  DEBUG ep 0x7f92e490b048: purge uct_ep[0]=0x55c2415402e0
  [1552583115.208388] [dgx05:35955:0]         ucp_ep.c:680  UCX  DEBUG ep 0x7f92e490b048: destroy uct_ep[0]=0x55c2415402e0
  [1552583115.209040] [dgx05:35955:0]      ib_device.c:644  UCX  DEBUG modify QP 0x2fe16 to state 0
  [1552583115.209063] [dgx05:35955:0]          rc_ep.c:211  UCX  DEBUG destroy rc ep 0x55c2415402e0
```

on the hang we just have 

```
  [1552583115.219809] [dgx05:35955:0]          ud_ep.c:1252 UCX  DEBUG ep 0x55c24315b0c0: disconnect
```

so probably hanging when closing the endpoint.

**Comment by Akshay-Venkatesh** [comment]
Fixed with 2b0594c

**Comment by Akshay-Venkatesh** [comment]
@TomAugspurger cleaned up the points where reader is added and removed in ucp module here:

https://gist.github.com/Akshay-Venkatesh/95cb4ba2c9d4684200557563869fd701#file-fix-send-recv-obj-hang-diff

Will make a pull request soon but these seems to fix the hangs.
